frontend/tests/unit/services/web3-provider/metamask-provider.spec.ts

```diff
@@ -0,0 +1,125 @@
+import * as providers from '@ethersproject/providers';
+import * as ethersUtils from 'ethers/lib/utils';
+
+import {
+  createMetaMaskProvider,
+  MetaMaskProvider,
+  onboardMetaMask,
+} from '@/services/web3-provider';
+import * as utils from '@/services/web3-provider/util-export';
+import { generateChain, getRandomEthereumAddress } from '~/utils/data_generators';
+import { MockedEip1193Provider, MockedWeb3Provider } from '~/utils/mocks/ethereum-provider';
+
+vi.mock('@/services/web3-provider/util-export');
+vi.mock('@ethersproject/providers');
+vi.mock('ethers/lib/utils');
+
+function mockWeb3Provider() {
+  const web3Provider = new MockedWeb3Provider();
+
+  Object.defineProperty(providers, 'Web3Provider', {
+    value: vi.fn().mockImplementation(() => web3Provider),
+  });
+
+  return web3Provider;
+}
+
+describe('metamask-provider', () => {
+  beforeEach(() => {
+    mockWeb3Provider();
+  });
+
+  describe('createMetaMaskProvider()', () => {
+    describe('when MetaMask is available', () => {
+      it('creates & initializes a MetaMask wallet provider', async () => {
+        Object.defineProperty(utils, 'detectEthereumProvider', {
+          value: vi.fn().mockResolvedValue(new MockedEip1193Provider({ isMetaMask: true })),
+        });
+
+        const chain = generateChain({ identifier: 2 });
+        const signerAddress = getRandomEthereumAddress();
+        const signer = 'fake-signer';
+
+        const web3Provider = mockWeb3Provider();
+        web3Provider.getNetwork = vi.fn().mockReturnValue({ chainId: chain.identifier });
+        web3Provider.listAccounts = vi.fn().mockReturnValue([signerAddress]);
+        web3Provider.getSigner = vi.fn().mockReturnValue(signer);
+
+        const metamaskProvider = await createMetaMaskProvider();
+
+        expect(metamaskProvider?.chainId.value).toBe(chain.identifier);
+        expect(metamaskProvider?.signer.value).toBe(signer);
+        expect(metamaskProvider?.signerAddress.value).toBe(signerAddress);
```

Comment:
Its because the `detectEthereumProvider` function, by design, can return undefined when a provider couldn't be detected.
We are using this function here: https://github.com/beamer-bridge/beamer/pull/1232/files#diff-a3c7eaee0e6a4178ecfef17aa1d1f14b5a923f0fc220bbfe55e0844bf56d872cL9

This function comes from this library: https://www.npmjs.com/package/@metamask/detect-provider
You can easily verify this by checking their usage docs:
```
<script src="https://unpkg.com/@metamask/detect-provider/dist/detect-provider.min.js"></script>
<script type="text/javascript">
  const provider = await detectEthereumProvider()

  if (provider) {
    // handle provider
  } else {
    // handle no provider
  }
</script>
```

So, in order to handle this case where this function will return `undefined` we had to set the type of our function: https://github.com/beamer-bridge/beamer/pull/1232/files#diff-a3c7eaee0e6a4178ecfef17aa1d1f14b5a923f0fc220bbfe55e0844bf56d872cL9
to return `Promise<MetaMaskProvider | undefined>` and therefore we have to use the `?.` everywhere so we are sure that we are not trying to access properties of an `undefined`.

Reply:
Ah, because TS is going to complain. I knew that it can be undefined, but basically if it is undefined the test will fail - which is fine. But TS complains that you are trying to access something that is might be undefined. Okaaaay.

Reply:
Exactly.